File: day33/houseRobber2.cpp
// 213. House Robber II
#include <iostream>
#include <vector>
using namespace std;

int maxSumOfNonAdjacent(vector<int> nums, int st, int end) {
  if(st == end ) return nums[st];
  
  int prev2 = 0, prev = nums[st];
  for (int i = st+1; i <= end; i++) {
    int take = nums[i] + (i==1? 0 : prev2);
    int notTake = 0 + prev;
    int curr = max(take, notTake);

    prev2 = prev;
    prev = curr;
  }
  return prev;
}

long long int houseRobber(vector<int>& nums) {
  if (nums.size() == 0 ) return 0;
  if (nums.size() == 1 ) return nums[0];

  return max(maxSumOfNonAdjacent(nums, 1, nums.size() - 1), maxSumOfNonAdjacent(nums, 0, nums.size() - 2));
}

int main() {
  vector<int> nums = {2, 1, 4, 9};
  cout << houseRobber(nums) << "\n";

  vector<int> nums2 = {2, 3, 2};
  cout << houseRobber(nums2) << "\n";

  vector<int> nums3 = {3, 2};
  cout << houseRobber(nums3) << "\n";

  return 0;
}

File: day112/HouseRobber2.java
// 213. House Robber II

public class HouseRobber2{
  private int[] dp;

  private int tabHelper(int start, int end, int[] nums){
    int n = end - start + 1;
    if (n == 1) return nums[start];

    dp = new int[n];
    dp[0] = nums[start];
    dp[1] = Math.max(nums[start], nums[start+1]);

    for(int i=2; i<n; ++i){
      dp[i] = Math.max(nums[start+i]+ dp[i-2], dp[i-1]);
    }
    return dp[n-1];
  }
  public int rob(int[] nums) {
    int n = nums.length;
    if(n==0)return 0;
    if(n==1) return nums[0];
    return Math.max(tabHelper(0, n-2, nums), tabHelper(1, n-1, nums));
  }


  private int tabHelper2(int start, int end, int[] nums){
    int n = end-start +1;
    if(n==1) return nums[start];

    int prev2 = nums[start];
    int prev = Math.max(nums[start], nums[start+1]);

    for(int i=2; i<n; ++i){
      int curr = Math.max(nums[start+i]+ prev2, prev);

      prev2 = prev;
      prev = curr;
    }
    return prev;
  }

  // Optimize Tabulation
  public int rob2(int[] nums) {
    int n = nums.length;
    if(n==0)return 0;
    if(n==1) return nums[0];
    return Math.max(tabHelper2(0, n-2, nums), tabHelper2(1, n-1, nums));
  }
  public static void main(String[] args) {
    HouseRobber2 hr = new HouseRobber2();
    System.out.println(hr.rob2(new int[] { 1,2,3,1 }));
  }
}